### PR TITLE
Rmc 134 log fulfilment requested event

### PR DESCRIPTION
**Pull request by hbrace**
# Motivation and Context

This logs an event in the case service when a fulfilment request is submitted.

Also fixes a defect where the event date time supplied in the inbound event is overwritten with the datetime of when it is processed by the case processor.

# What has changed

Fulfilment request objects and processor created along with unit and integration tests.

# How to test?

Build this image and the [case-api branch of the same name](https://github.com/ONSdigital/census-rm-case-api/pull/19) and run against this [acceptance test branch](https://github.com/ONSdigital/census-rm-acceptance-tests/pull/95).

# Links

https://trello.com/c/0cFn8BaV/990-rmc-134a-log-fulfilment-requested-event-5
https://trello.com/c/L9OTCN0S/1013-case-processor-log-receipt-and-refusal-event-date-time-is-overwritten-with-current-date-time

# Screenshots (if appropriate):

### CODE REVIEW

**Comment by hbrace** [comment]
[RMC-134(A) Log Fulfilment requested event (5)](https://trello.com/c/0cFn8BaV/990-rmc-134a-log-fulfilment-requested-event-5)

File: src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    Optional<Case> cazeResult = caseRepository.findByCaseId(UUID.fromString(request.getCaseId()));
+
+    if (cazeResult.isEmpty()) {
+      log.error(CASE_NOT_FOUND_ERROR);
+      throw new RuntimeException(String.format("Case ID '%s' not found!", request.getCaseId()));
+    }
+
+    Case caze = cazeResult.get();
+
+    int caseRef = caze.getCaseRef();
```
Comment (ryangrundy7) [inline src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java]: Looks like you aren't using this variable

File: src/test/java/uk/gov/ons/census/casesvc/testutil/DataUtils.java [hunk 2]
```diff
@@ -78,6 +79,28 @@ public static ResponseManagementEvent getTestResponseManagementRefusalEvent() {
     return managementEvent;
   }
 
+  public static ResponseManagementEvent getTestResponseManagementFulfilmentRequestedEvent() {
+    ResponseManagementEvent managementEvent = getTestResponseManagementEvent();
+
+    EventDTO event = managementEvent.getEvent();
+    event.setType(EventType.REFUSAL_RECEIVED);
```
Comment (ryangrundy7) [inline src/test/java/uk/gov/ons/census/casesvc/testutil/DataUtils.java]: This should be set to `FULFILMENT_REQUESTED`

File: src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java (hunk 1 above)
Comment (hbrace) [reply src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java]: Removed 👍

File: src/test/java/uk/gov/ons/census/casesvc/testutil/DataUtils.java (hunk 2 above)
Comment (hbrace) [reply src/test/java/uk/gov/ons/census/casesvc/testutil/DataUtils.java]: I should pay more attention when copy and pasting

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java [hunk 3]
```diff
@@ -71,6 +73,30 @@ public void logRefusalEvent(
...
+      String eventDescription,
+      EventType eventType,
+      FulfilmentRequestDTO payload,
+      EventDTO event) {
+
+    Event loggedEvent = new Event();
+    loggedEvent.setCaze(caze);
+    loggedEvent.setId(UUID.randomUUID());
+    loggedEvent.setCaseId(caseId);
+    loggedEvent.setEventDate(OffsetDateTime.now());
```
Comment (liamtoozer) [inline src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: Is it worth using the date from the event itself here?

File: src/test/java/uk/gov/ons/census/casesvc/testutil/DataUtils.java [hunk 4]
```diff
@@ -82,6 +83,28 @@ public static ResponseManagementEvent getTestResponseManagementRefusalEvent() {
...
+
+    PayloadDTO payload = managementEvent.getPayload();
+    payload.setUac(null);
+    payload.setCollectionCase(null);
+    payload.setReceipt(null);
+    payload.setPrintCaseSelected(null);
+    payload.setRefusal(null);
+
+    FulfilmentRequestDTO fulfilmentRequest = payload.getFulfilmentRequest();
+    fulfilmentRequest.setCaseId(null);
```
Comment (ryangrundy7) [inline src/test/java/uk/gov/ons/census/casesvc/testutil/DataUtils.java]: This might be picky but shouldn't caseId never be null?

**Comment by liamtoozer** [comment]
[Case Processor - Log receipt and refusal event date/time is overwritten with current date/time](https://trello.com/c/L9OTCN0S/1013-case-processor-log-receipt-and-refusal-event-date-time-is-overwritten-with-current-date-time)

**Comment by liamtoozer** [comment]
[RMC-197 Handle Refusal Received Event [REV13] (8)](https://trello.com/c/29ZsK7ci/825-rmc-197-handle-refusal-received-event-rev13-8)

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java (hunk 3 above)
Comment (davidmort) [reply src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: yep agreed, this should be the date contained in the event passed to us

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java [hunk 5]
```diff
@@ -71,6 +73,30 @@ public void logRefusalEvent(
...
 
+  public void logFulfilmentRequestedEvent(
+      Case caze,
+      UUID caseId,
+      String eventDescription,
+      EventType eventType,
+      FulfilmentRequestDTO payload,
+      EventDTO event) {
+
+    Event loggedEvent = new Event();
```
Comment (davidmort) [inline src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: is there a reason we aren't called "logEvent" like we do for the logReceiptEvent and logRefusalEvent methods?

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java [hunk 6]
```diff
@@ -71,6 +73,30 @@ public void logRefusalEvent(
...
+    loggedEvent.setCaze(caze);
+    loggedEvent.setId(UUID.randomUUID());
+    loggedEvent.setCaseId(caseId);
+    loggedEvent.setEventDate(OffsetDateTime.now());
+    loggedEvent.setEventDescription(eventDescription);
+    loggedEvent.setEventType(eventType);
+    loggedEvent.setEventPayload(convertObjectToJson(payload));
+    loggedEvent.setEventChannel(event.getChannel());
+    loggedEvent.setEventSource(event.getSource());
+    loggedEvent.setEventTransactionId(UUID.randomUUID());
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: shouldn't transaction id be from event?

File: src/main/java/uk/gov/ons/census/casesvc/config/QueueSetterUpper.java [hunk 7]
```diff
@@ -42,6 +42,9 @@
   @Value("${queueconfig.refusal-routing-key}")
   private String caseProcessorRefusalRoutingKeyCase;
 
+  @Value("${queueconfig.fulfilment-routing-key}")
+  private String caseProcessorFulfilmentRoutingKeyCase;
```
Comment (liamtoozer) [inline src/main/java/uk/gov/ons/census/casesvc/config/QueueSetterUpper.java]: Tiny (picky!) comment but is it worth removing `caseProcessor` prefix as it lives in this service anyway? I guess the refusal bits are configured in the same way though 🤔

File: src/main/java/uk/gov/ons/census/casesvc/model/dto/FulfilmentRequestDTO.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+public class FulfilmentRequestDTO {
+
+  @JsonInclude(Include.NON_NULL)
+  @JsonProperty("caseId")
+  private String caseId;
+
+  private String fulfilmentCode;
+
+  private String individualCaseId;
+}
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/model/dto/FulfilmentRequestDTO.java]: event dictionary shows "address" & "contact" as part of a "fulfilmentRequest" - presumably, these are left out deliberately as with RefusalReceived?

File: src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+@Service
+public class FulfilmentRequestProcessor {
+
+  private static final Logger log = LoggerFactory.getLogger(FulfilmentRequestProcessor.class);
+
+  private static final String CASE_NOT_FOUND_ERROR = "Case not found error";
+
+  private final CaseRepository caseRepository;
+  private final EventLogger eventLogger;
+  private final EventRepository eventRepository;
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java]: redundant

File: src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+      log.error(CASE_NOT_FOUND_ERROR);
+      throw new RuntimeException(String.format("Case ID '%s' not found!", request.getCaseId()));
+    }
+
+    Case caze = cazeResult.get();
+
+    eventLogger.logFulfilmentRequestedEvent(
+        caze,
+        UUID.fromString(request.getCaseId()),
+        "Fulfilment Request Received",
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java]: for consistency, use "private static final String FULFILMENT_REQUEST_RECEIVED = "Fulfilment Request Received";"

File: src/test/java/uk/gov/ons/census/casesvc/messaging/FulfilmentRequestReceiverTest.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+import org.junit.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import uk.gov.ons.census.casesvc.model.dto.ResponseManagementEvent;
+import uk.gov.ons.census.casesvc.service.FulfilmentRequestProcessor;
+
+public class FulfilmentRequestReceiverTest {
+
+  @Mock private FulfilmentRequestProcessor fulfilmentRequestProcessor;
```
Comment (AdeHarrison) [inline src/test/java/uk/gov/ons/census/casesvc/messaging/FulfilmentRequestReceiverTest.java]: redundant

File: src/test/java/uk/gov/ons/census/casesvc/messaging/FulfilmentRequestReceiverTest.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+  @InjectMocks FulfilmentRequestReceiver underTest;
```
Comment (AdeHarrison) [inline src/test/java/uk/gov/ons/census/casesvc/messaging/FulfilmentRequestReceiverTest.java]: redundant

File: src/test/java/uk/gov/ons/census/casesvc/messaging/FulfilmentRequestReceiverIT.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+@ContextConfiguration
+@ActiveProfiles("test")
+@SpringBootTest
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+@RunWith(SpringJUnit4ClassRunner.class)
+public class FulfilmentRequestReceiverIT {
+
+  private static final UUID TEST_CASE_ID = UUID.randomUUID();
+  private static final String TEST_FULFILMENT_CODE = "P_IC_ICL1";
+  private static final EasyRandom easyRandom = new EasyRandom();
```
Comment (AdeHarrison) [inline src/test/java/uk/gov/ons/census/casesvc/messaging/FulfilmentRequestReceiverIT.java]: redundant

File: src/test/java/uk/gov/ons/census/casesvc/messaging/FulfilmentRequestReceiverIT.java [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+  @Transactional
+  public void setUp() {
+    rabbitQueueHelper.purgeQueue(inboundQueue);
+    eventRepository.deleteAllInBatch();
+    uacQidLinkRepository.deleteAllInBatch();
+    caseRepository.deleteAllInBatch();
+  }
+
+  @Test
+  public void testFulfilmentRequestLogged() throws IOException, InterruptedException {
```
Comment (AdeHarrison) [inline src/test/java/uk/gov/ons/census/casesvc/messaging/FulfilmentRequestReceiverIT.java]: IOException is never thrown

File: src/main/java/uk/gov/ons/census/casesvc/model/dto/FulfilmentRequestDTO.java (hunk 8 above)
Comment (hbrace) [reply src/main/java/uk/gov/ons/census/casesvc/model/dto/FulfilmentRequestDTO.java]: Yep deliberately left out

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java (hunk 6 above)
Comment (hbrace) [reply src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: Changes made in my last commit to address this, let me know what you think

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java [hunk 15]
```diff
@@ -34,59 +37,65 @@ public void logEvent(
...
-        eventType,
-        convertObjectToJson(payload),
-        event,
-        eventMetaDataDateTime);
+      EventDTO event) {
+
+    logEvent(uacQidLink, eventDescription, eventType, convertObjectToJson(payload), event);
+  }
+
+  public void logFulfilmentRequestedEvent(
```
Comment (liamtoozer) [inline src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: Think this method needs a direct test as well, similar to [this one for receipting](https://github.com/ONSdigital/census-rm-case-processor/blob/0a08da43334c87ed1795734a21bf51ec235140a2/src/test/java/uk/gov/ons/census/casesvc/logging/EventLoggerTest.java#L89)?

File: src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+
+  @Mock private CaseRepository caseRepository;
+
+  @Mock private EventLogger eventLogger;
+
+  @InjectMocks FulfilmentRequestProcessor underTest;
+
+  @Test
+  public void testGoodFulfilmentRequest() {
+    ResponseManagementEvent managementEvent = getTestResponseManagementEvent();
```
Comment (liamtoozer) [inline src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java]: Still works fine, but might be worth using the tailored test object for fulfilment here?

```suggestion
    ResponseManagementEvent managementEvent = getTestResponseManagementFulfilmentRequestedEvent();
```

File: src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+            "Fulfilment Request Received",
+            FULFILMENT_REQUESTED,
+            expectedFulfilmentRequest,
+            managementEvent.getEvent());
+  }
+
+  @Test(expected = RuntimeException.class)
+  public void testCaseIdNotFound() {
+    // GIVEN
+    ResponseManagementEvent managementEvent = getTestResponseManagementEvent();
```
Comment (liamtoozer) [inline src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java]: Still works fine, but might be worth using the tailored test object for fulfilment here?

```suggestion
    ResponseManagementEvent managementEvent = getTestResponseManagementFulfilmentRequestedEvent();
```

File: src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    } catch (RuntimeException re) {
+      // THEN
+      assertThat(re.getMessage()).isEqualTo(expectedErrorMessage);
+      throw re;
+    }
+  }
+
+  @Test(expected = RuntimeException.class)
+  public void testNullDateTime() {
+    ResponseManagementEvent managementEvent = getTestResponseManagementEvent();
```
Comment (liamtoozer) [inline src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java]: Still works fine, but might be worth using the tailored test object for fulfilment here?

```suggestion
    ResponseManagementEvent managementEvent = getTestResponseManagementFulfilmentRequestedEvent();
```

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java (hunk 15 above)
Comment (hbrace) [reply src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: Test added 👍

File: src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java (hunk 16 above)
Comment (hbrace) [reply src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java]: Receipt and refusal tests both use the generic method call so just following the consistency!

File: src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java (hunk 17 above)
Comment (hbrace) [reply src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java]: Receipt and refusal tests both use the generic method call so just following the consistency!

File: src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java (hunk 18 above)
Comment (hbrace) [reply src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java]: Receipt and refusal tests both use the generic method call so just following the consistency!

File: src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java (hunk 6 above)
Comment (AdeHarrison) [reply src/main/java/uk/gov/ons/census/casesvc/logging/EventLogger.java]: that's the best you can do with confirmation from dave

File: src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+  private static final String CASE_NOT_FOUND_ERROR = "Case not found error";
+  private static final String DATETIME_NOT_PRESENT = "Date time not in event error";
+  private static final String FULFILMENT_REQUEST_RECEIVED = "Fulfilment Request Received";
+
+  private final CaseRepository caseRepository;
+  private final EventLogger eventLogger;
+
+  public FulfilmentRequestProcessor(
+      CaseRepository caseRepository, EventLogger eventLogger, EventRepository eventRepository) {
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java]: eventRepository not used

File: src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+import uk.gov.ons.census.casesvc.model.dto.EventDTO;
+import uk.gov.ons.census.casesvc.model.dto.FulfilmentRequestDTO;
+import uk.gov.ons.census.casesvc.model.dto.ResponseManagementEvent;
+import uk.gov.ons.census.casesvc.model.entity.Case;
+import uk.gov.ons.census.casesvc.model.repository.CaseRepository;
+
+@RunWith(MockitoJUnitRunner.class)
+public class FulfilmentRequestProcessorTest {
+
+  @Mock private CaseProcessor caseProcessor;
```
Comment (AdeHarrison) [inline src/test/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessorTest.java]: not used

File: src/main/java/uk/gov/ons/census/casesvc/model/dto/FulfilmentRequestDTO.java [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+public class FulfilmentRequestDTO {
+
+  @JsonInclude(Include.NON_NULL)
+  @JsonProperty("caseId")
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/model/dto/FulfilmentRequestDTO.java]: don't need this (appreciate it's in receipt & refusal which is redundant there too)

File: src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+      log.error(DATETIME_NOT_PRESENT);
+      throw new RuntimeException(
+          String.format("Date time not found in fulfilment request event for case '%s", caseId));
+    }
+
+    Case caze = cazeResult.get();
+
+    eventLogger.logFulfilmentRequestedEvent(
+        caze,
+        UUID.fromString(fulfilmentRequestPayload.getCaseId()),
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java]: really picky but could use caseId from line 41

File: src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+        fulfilmentRequest.getPayload().getFulfilmentRequest();
+
+    String caseId = fulfilmentRequestPayload.getCaseId();
+
+    Optional<Case> cazeResult = caseRepository.findByCaseId(UUID.fromString(caseId));
+
+    if (cazeResult.isEmpty()) {
+      log.error(CASE_NOT_FOUND_ERROR);
+      throw new RuntimeException(
+          String.format("Case ID '%s' not found!", fulfilmentRequestPayload.getCaseId()));
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java]: picky again, use caseId from line 41

File: src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    if (event.getDateTime() == null) {
+      log.error(DATETIME_NOT_PRESENT);
+      throw new RuntimeException(
+          String.format("Date time not found in fulfilment request event for case '%s", caseId));
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java]: picky, picky "Case ID" in error message

File: src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+    Case caze = cazeResult.get();
+
+    eventLogger.logFulfilmentRequestedEvent(
+        caze,
+        UUID.fromString(fulfilmentRequestPayload.getCaseId()),
+        event.getDateTime(),
+        FULFILMENT_REQUEST_RECEIVED,
+        FULFILMENT_REQUESTED,
+        fulfilmentRequestPayload,
+        fulfilmentRequest.getEvent());
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/service/FulfilmentRequestProcessor.java]: event already extracted in line 37?

File: src/main/java/uk/gov/ons/census/casesvc/model/dto/FulfilmentRequestDTO.java [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+  private String caseId;
+
+  private String fulfilmentCode;
+
+  private String individualCaseId;
```
Comment (AdeHarrison) [inline src/main/java/uk/gov/ons/census/casesvc/model/dto/FulfilmentRequestDTO.java]: could add @JsonInclude(Include.NON_NULL) above here to suppress null in event_payload field

**Pull request merged**